### DISCUSSION TITLE
Set state inside zod SuperRefine or in a hook called by zod

### DISCUSSION THREAD

**Discussion by ithilbor** [colinhacks/zod#2149] [Q&A] [answered]
Hi, is there a way to set a state into the superRefine function or into a hook called by the zod schema? 
By now when the zod schema calls the hook that set the state I'm getting the error: 

`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'length')
    at n (zod.ts:15:1)
    at zod.ts:77:1
    at async _executeSchema (createFormControl.ts:381:1)
    at async Object.onChange (createFormControl.ts:718:1)`
    
Is that normal? Is there a way to do this? Thanks in advance!

Zod version: 3.21.2

PS: Let me know if you need a codesandbox

**Comment by JacobWeisenburger** [comment]
> PS: Let me know if you need a codesandbox

When asking any question that involves code, I think it's safe to assume that if you are able to provide a reproducible code example, then you should do so. This saves a lot of time for someone that wants to answer your question. Also it shows that you care enough to put in effort to help get you the info you are asking for. Generally, the amount of effort you are willing to put into asking a question, will be the same amount of effort that someone will put into answering your question.

**Reply by ithilbor** [reply]
You're perfectly right @JacobWeisenburger and I fully agree with you, but I was in the working hours and I couldn't do it until now so I submitted the "short version" of the question without the sandbox.
Now that I've "finished" my working hours I had the time to do the sandbox, so here it is:  https://codesandbox.io/s/test-setstate-with-zod-30vlxr

In order to see the error that I've posted before compile and submit the form.

**Reply by JacobWeisenburger** [reply]
The sandbox you gave is very complex and has other things outside of zod. Can you provide an example that only involves zod?

**Reply by ithilbor** [reply]
I think I can do It, now I'm going to eat something for dinner and after I'll submit the new example (I'll be here in minutes)

**Comment by JacobWeisenburger** [comment]
after looking into this some more I believe the problem is you are calling a hook in a nested function.

[Only Call Hooks at the Top Level](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by ithilbor** [reply]
Here I am! The simplified version of the previous sandbox is this one: https://codesandbox.io/s/test-setstate-with-zod-forked-i68ny1?file=/src/App.tsx

By simplify the sandbox and using the parse I noticed that zod doesn't receives the value and that's why it gives the error on string length

**Reply by ithilbor** [reply]
PS: I will also sponsor you with buy me a sandwich!

**Reply by JacobWeisenburger** [reply]
Thanks so much for the sponsorship! 🙏

**Reply by ithilbor** [reply]
As you said the error can be caused because it is not called at the top level so the question now can be how can i trigger a zod error in my component if the axios call fails?

**Comment by JacobWeisenburger** [comment, marked as answer]
> Here I am! The simplified version of the previous sandbox is this one: https://codesandbox.io/s/test-setstate-with-zod-forked-i68ny1?file=/src/App.tsx
> 
> By simplify the sandbox and using the parse I noticed that zod doesn't receives the value and that's why it gives the error on string length

I see a few things going on here:
1. You are calling a hook in a nested function
    - [Only Call Hooks at the Top Level](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)
2. You are using `parse` on an async super refine. You need to use `parseAsync`
3. the axios request is throwing an error, but you are not handling it in your try/catch block.
4. You misspelled 'something' as 'someting'

---

I fixed these problems and it seems to be working here:
https://codesandbox.io/s/test-setstate-with-zod-forked-mohx4z?file=/src/App.tsx

**Reply by ithilbor** [reply]
Thanks @JacobWeisenburger I wrote the code to much faster, the demo now works but I can't do what I want. In the first sandbox that I've posted here: https://codesandbox.io/s/test-setstate-with-zod-30vlxr

What I was trying to do was: if the axios call goes well set a state with the returned date in order to make them available to the `Login` component  this way zod makes all the validation and my component can still get the data of the axios call.
But as I saw I can not set the data into a state so what I can do is to set the state in the `Login`. component and with react hook form set an error if the axios call doesn't go well. This type of solution has brought me this problem:

1. The error goes away each time the zod validation is triggered (even when the user submit the forms)

If you have a suggestion to do this I will listen to you, thanks again @JacobWeisenburger! Maybe lately I'll update the codesandbox with the problem explained here (the disappear of the error)

**Reply by JacobWeisenburger** [reply]
I've tried playing around with this, but it is a very complex problem, so I wasn't able to find a good solution without rewriting a lot of your code. Also this involves a lot more then just zod, so unfortunately, I'm not able to help anymore. I'm more than happy to help with problems that only involve zod.

**Reply by ithilbor** [reply]
Don't you worry any more @JacobWeisenburger , thank you so much for your help, I'll work on this problem, now I think I've found a solution to get the behaviour that I want but it need to be cleared because the code can be difficult to read and understand now! Thanks again, wish you a happy and nice day!
